## Установка pandas

Установим библиотеку pandas:

In [ ]:
!pip install pandas

и импортируем ее:

In [41]:
import pandas as pd

## 1. Объекты Series и DataFrame

Объект `Series` – одномерные массив, состоящий из элементов и их индексов:

In [42]:
series_example = pd.Series(['a', 48, 'd', 'hello115'])
series_example

0           a
1          48
2           d
3    hello115
dtype: object

Здесь:  
`0`, `1`, `2`, `3` – индексы.  
`a`, `48`, `d`, `hello115` – элементы.

Объект `DataFrame` – многомерный массив, как таблица в Excel.

In [43]:
df = pd.DataFrame({
     'telephone number': [123456789, 123679, 15645445, 564645546],
     'name': ['John', 'Jim', 'Kotowski', 'Lebowski'],
})

df

telephone number      name
0         123456789      John
1            123679       Jim
2          15645445  Kotowski
3         564645546  Lebowski

Здесь:  
`telephone number`, `name` – названия столбцов с соответствующими значениями `123456789 ...`, `John ...`.

## 2. Импорт файлов и просмотр содержимого

Импортируем датасет ([оригинал](https://www.kaggle.com/dhruvildave/google-trends-dataset)) с популярными поисковыми запросами Google с 2000 по 2020 год. При импорте csv-файла все данные сохраняются в датафрейм `df`:

In [44]:
df = pd.read_csv('https://raw.githubusercontent.com/tttdddnet/Python-Data-Journalism/main/part1/trends.csv')

Аналогично есть команды для импорта других файлов:  

`read_excel()` – импорт excel-файла.  
`read_sql()` – импорт базы данных SQL.  
`read_json(json_string)` – импорт json-файла.  
`read_html(url)` – создает из HTML-таблицы список объектов DataFrame.  

Выведем содержимое датафрейма на экран:

In [45]:
print(df)

      location  year         category  rank                query
0       Global  2001  Consumer Brands     1                Nokia
1       Global  2001  Consumer Brands     2                 Sony
2       Global  2001  Consumer Brands     3                  BMW
3       Global  2001  Consumer Brands     4                 Palm
4       Global  2001  Consumer Brands     5                Adobe
...        ...   ...              ...   ...                  ...
26950  Vietnam  2020           Là Gì?     1   Virus Corona là gì
26951  Vietnam  2020           Là Gì?     2  Miễn thị thực là gì
26952  Vietnam  2020           Là Gì?     3    Đầu cắt moi là gì
26953  Vietnam  2020           Là Gì?     4  Bệnh bạch hầu là gì
26954  Vietnam  2020           Là Gì?     5       Đông Lào là gì

[26955 rows x 5 columns]


Отображаются первые и последние пять строчек.

Чаще всего значения в датасетах разделяются запятой (рис.). Чтобы задать другой разделитель добавим параметр `sep=""`. По умолчанию разделитель между столбцами – запятая.
В нашем случае получаем:

In [46]:
df = pd.read_csv('https://raw.githubusercontent.com/tttdddnet/Python-Data-Journalism/main/part1/trends.csv', sep=",")
df

location  year         category  rank                query
0       Global  2001  Consumer Brands     1                Nokia
1       Global  2001  Consumer Brands     2                 Sony
2       Global  2001  Consumer Brands     3                  BMW
3       Global  2001  Consumer Brands     4                 Palm
4       Global  2001  Consumer Brands     5                Adobe
...        ...   ...              ...   ...                  ...
26950  Vietnam  2020           Là Gì?     1   Virus Corona là gì
26951  Vietnam  2020           Là Gì?     2  Miễn thị thực là gì
26952  Vietnam  2020           Là Gì?     3    Đầu cắt moi là gì
26953  Vietnam  2020           Là Gì?     4  Bệnh bạch hầu là gì
26954  Vietnam  2020           Là Gì?     5       Đông Lào là gì

[26955 rows x 5 columns]

Посмотрим первые пять строчек датафрейма с помощью `df.head(5)`:

In [47]:
df.head(5)

location  year         category  rank  query
0   Global  2001  Consumer Brands     1  Nokia
1   Global  2001  Consumer Brands     2   Sony
2   Global  2001  Consumer Brands     3    BMW
3   Global  2001  Consumer Brands     4   Palm
4   Global  2001  Consumer Brands     5  Adobe

последние 5 строчек:

In [48]:
df.tail(5)

location  year category  rank                query
26950  Vietnam  2020   Là Gì?     1   Virus Corona là gì
26951  Vietnam  2020   Là Gì?     2  Miễn thị thực là gì
26952  Vietnam  2020   Là Gì?     3    Đầu cắt moi là gì
26953  Vietnam  2020   Là Gì?     4  Bệnh bạch hầu là gì
26954  Vietnam  2020   Là Gì?     5       Đông Lào là gì

и пять случайных строчек:

In [49]:
df.sample(5)

location  year                  category  rank  \
8985       Indonesia  2014  Pertandingan Piala Dunia     1   
13000  United States  2015               NBA Players     1   
25082          India  2020                   Near me     3   
3645         Hungary  2012                    Recipe     1   
22839    Netherlands  2019                  Sporters     5   

                       query  
8985     Jerman Vs Argentina  
13000          Stephen Curry  
25082  Crackers shop near me  
3645       Csirkemell Recept  
22839                 Neymar

Узнаем количество строчек и столбцов:

In [50]:
df.shape

(26955, 5)

Здесь:  
26955 – количество строчек.  
5 – количество столбцов.

Выведем на экран содержимое датафрейма:

In [51]:
df.info

<bound method DataFrame.info of       location  year         category  rank                query
0       Global  2001  Consumer Brands     1                Nokia
1       Global  2001  Consumer Brands     2                 Sony
2       Global  2001  Consumer Brands     3                  BMW
3       Global  2001  Consumer Brands     4                 Palm
4       Global  2001  Consumer Brands     5                Adobe
...        ...   ...              ...   ...                  ...
26950  Vietnam  2020           Là Gì?     1   Virus Corona là gì
26951  Vietnam  2020           Là Gì?     2  Miễn thị thực là gì
26952  Vietnam  2020           Là Gì?     3    Đầu cắt moi là gì
26953  Vietnam  2020           Là Gì?     4  Bệnh bạch hầu là gì
26954  Vietnam  2020           Là Gì?     5       Đông Lào là gì

[26955 rows x 5 columns]>

и использованной памяти:

In [52]:
df.info(memory_usage=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26955 entries, 0 to 26954
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   location  26955 non-null  object
 1   year      26955 non-null  int64 
 2   category  26955 non-null  object
 3   rank      26955 non-null  int64 
 4   query     26955 non-null  object
dtypes: int64(2), object(3)
memory usage: 1.0+ MB


Иногда при импорте данных числа принимают тип `объект` и при обработке данных возникает ошибка. Поэтому проверим какие типы данных находятся в каждом столбце:

In [53]:
df.dtypes

location    object
year         int64
category    object
rank         int64
query       object
dtype: object

Здесь:  
`object` – строчка.  
`int64` – целое число.  

Часто встречаемые типы данных:  
`float64` – вещественное число.  
`datetime64` – дата и время.  
`bool` – значения `True` или `False`.  


Узнаем названия всех столбцов:

In [54]:
df.columns

Index(['location', 'year', 'category', 'rank', 'query'], dtype='object')

Выведем значения только столбца `query`:

In [55]:
df['query']

0                      Nokia
1                       Sony
2                        BMW
3                       Palm
4                      Adobe
                ...         
26950     Virus Corona là gì
26951    Miễn thị thực là gì
26952      Đầu cắt moi là gì
26953    Bệnh bạch hầu là gì
26954         Đông Lào là gì
Name: query, Length: 26955, dtype: object

и значения нескольких столбцов:

In [56]:
df[['location', 'category']]

location         category
0       Global  Consumer Brands
1       Global  Consumer Brands
2       Global  Consumer Brands
3       Global  Consumer Brands
4       Global  Consumer Brands
...        ...              ...
26950  Vietnam           Là Gì?
26951  Vietnam           Là Gì?
26952  Vietnam           Là Gì?
26953  Vietnam           Là Gì?
26954  Vietnam           Là Gì?

[26955 rows x 2 columns]

Через срез отобразим первые двадцать значений столбца `query`:

In [57]:
df['query'][0:20]

0                 Nokia
1                  Sony
2                   BMW
3                  Palm
4                 Adobe
5           Nostradamus
6       Osama bin Laden
7                Eminem
8       Michael Jackson
9          Howard Stern
10         Harry Potter
11    Lord of the Rings
12        Final Fantasy
13          Tomb Raider
14                Shrek
15             Morpheus
16              Napster
17             Gnutella
18                Kazaa
19          Audiogalaxy
Name: query, dtype: object

В нашем датасете индексы являются числами. Это не очень удобно для понимания, как работает `loc` и `iloc`, поэтому создадим небольшой датафрейм с элементами типа `str` в виде индексов:

In [58]:
df = pd.DataFrame([[1, 2], [3, 4], [5, 6]],
     index=['phone', 'telegram', 'letter'],
     columns=['max_speed', 'cost'])
df

max_speed  cost
phone             1     2
telegram          3     4
letter            5     6

`loc` выдает строки и/или столбцы по нечисловому значению индекса:

In [59]:
df.loc['phone']

max_speed    1
cost         2
Name: phone, dtype: int64

`iloc` получает строки и/или столбцы по числовому индексу:

In [60]:
df.iloc[0]

max_speed    1
cost         2
Name: phone, dtype: int64

Вернемся к основному датафрейму:

In [61]:
df = pd.read_csv('https://raw.githubusercontent.com/tttdddnet/Python-Data-Journalism/main/part1/trends.csv')

Выведем первые пять строчек с помощью `iloc`:

In [62]:
df.iloc[0:5]

location  year         category  rank  query
0   Global  2001  Consumer Brands     1  Nokia
1   Global  2001  Consumer Brands     2   Sony
2   Global  2001  Consumer Brands     3    BMW
3   Global  2001  Consumer Brands     4   Palm
4   Global  2001  Consumer Brands     5  Adobe

Узнаем какой элемент записан в третьей строчке и пятом столбце:

In [63]:
df.iloc[2,4]

'BMW'

Выведем значения всех строчек с помощью метода `iterrows()`:

In [ ]:
for index, row in df.iterrows():
    print(row)

 и с помощью `iloc()`:

In [ ]:
for i in range(len(df.index)):
    print(df.iloc[i])

Отобразим все элементы столбца `query`:

In [ ]:
for index, row in df.iterrows():
    print(row['query'])

иначе:

In [ ]:
for i in range(len(df.index)):
    print(df.iloc[i]['query'])

Выведем строчки со значением `Consumer Brands` в столбце `Сategory`

In [68]:
df.loc[df['category'] == "Consumer Brands"]

location  year         category  rank       query
0     Global  2001  Consumer Brands     1       Nokia
1     Global  2001  Consumer Brands     2        Sony
2     Global  2001  Consumer Brands     3         BMW
3     Global  2001  Consumer Brands     4        Palm
4     Global  2001  Consumer Brands     5       Adobe
65    Global  2002  Consumer Brands     1     Ferrari
66    Global  2002  Consumer Brands     2        Sony
67    Global  2002  Consumer Brands     3       Nokia
68    Global  2002  Consumer Brands     4      Disney
69    Global  2002  Consumer Brands     5        IKEA
175   Global  2003  Consumer Brands     1     Ferrari
176   Global  2003  Consumer Brands     2        Sony
177   Global  2003  Consumer Brands     3         BMW
178   Global  2003  Consumer Brands     4      Disney
179   Global  2003  Consumer Brands     5     Ryanair
340   Global  2004  Consumer Brands     1        eBay
341   Global  2004  Consumer Brands     2     Walmart
342   Global  2004  Consumer Brands     3    MapQuest
343   Global  2004  Consumer Brands     4      Amazon
344   Global  2004  Consumer Brands     5  Home Depot

## 3. Операции над строчками и столбцами

### 3.1. Объединение датафреймов

Создадим два датафрейма `df1` и `df2` и добавим столбцы из второго датафрейма в первый с помощью `concat()`. При этом индексы должны совпадать, иначе мы получим `NaN`:

In [69]:
df1 = pd.DataFrame([[1, 2], [3, 4], [5, 6]],
     index=['phone', 'telegram', 'letter'],
     columns=['max_speed', 'cost'])

df2 = pd.DataFrame([[7, 8], [9, 10], [11, 12]],
     index=['phone', 'telegram', 'letter'],
     columns=['average speed', 'min_speed'])

df1 = pd.concat([df1, df2], axis=1)
df1

max_speed  cost  average speed  min_speed
phone             1     2              7          8
telegram          3     4              9         10
letter            5     6             11         12

Здесь:  
`df1` и `df2` – первый и второй датафрейм соответственно.

Добавим строчки датафрейма `df2` к `df1`. Названия столбцов должны совпадать, иначе получим `NaN`:

In [70]:
df1 = pd.DataFrame([[1, 2], [3, 4], [5, 6]],
     index=['phone', 'telegram', 'letter'],
     columns=['max_speed', 'cost'])

df2 = pd.DataFrame([[7, 8], [9, 10], [11, 12]],
     index=['smartphone', 'birdmail', 'email'],
     columns=['max_speed', 'cost'])

df1 = pd.concat([df1, df2], axis=0)
df1

max_speed  cost
phone               1     2
telegram            3     4
letter              5     6
smartphone          7     8
birdmail            9    10
email              11    12

Добавим строчки методом `append()`:

In [71]:
df1 = pd.DataFrame([[1, 2], [3, 4], [5, 6]],
     index=['phone', 'telegram', 'letter'],
     columns=['max_speed', 'cost'])

df2 = pd.DataFrame([[7, 8], [9, 10], [11, 12]],
     index=['smartphone', 'birdmail', 'email'],
     columns=['max_speed', 'cost'])

df1 = df1.append(df2)
df1

max_speed  cost
phone               1     2
telegram            3     4
letter              5     6
smartphone          7     8
birdmail            9    10
email              11    12

### 3.2. Добавление строчек и столбцов

Для добавления строчки создадим словарь `new_row` с названиям столбцов ввиде ключей:

In [72]:
df = pd.read_csv('https://raw.githubusercontent.com/tttdddnet/Python-Data-Journalism/main/part1/trends.csv')
new_row = {'location': 'Russia', 'year': 2021, 'category': 'Music', 'rank': 1, 'query': 'Artist Name'}
df = df.append(new_row, ignore_index=True)
df.tail(5)

location  year category  rank                query
26951  Vietnam  2020   Là Gì?     2  Miễn thị thực là gì
26952  Vietnam  2020   Là Gì?     3    Đầu cắt moi là gì
26953  Vietnam  2020   Là Gì?     4  Bệnh bạch hầu là gì
26954  Vietnam  2020   Là Gì?     5       Đông Lào là gì
26955   Russia  2021    Music     1          Artist Name

Чтобы добавить строчку в определенную позицию используем метод `loc`:

In [73]:
df.loc[2] = ['Russia', 2021, 'Music', 1, 'Artist Name']
df.head(5)

location  year         category  rank        query
0   Global  2001  Consumer Brands     1        Nokia
1   Global  2001  Consumer Brands     2         Sony
2   Russia  2021            Music     1  Artist Name
3   Global  2001  Consumer Brands     4         Palm
4   Global  2001  Consumer Brands     5        Adobe

Здесь:  
`loc[2]` – вставляем строчку на место третьей строчки.

Следующая запись добавит новый столбец `new_column` справа и установит все строки на значение `new`:

In [74]:
df = pd.read_csv('https://raw.githubusercontent.com/tttdddnet/Python-Data-Journalism/main/part1/trends.csv')
df['new_column'] = 'new'
df.tail(5)

location  year category  rank                query new_column
26950  Vietnam  2020   Là Gì?     1   Virus Corona là gì        new
26951  Vietnam  2020   Là Gì?     2  Miễn thị thực là gì        new
26952  Vietnam  2020   Là Gì?     3    Đầu cắt moi là gì        new
26953  Vietnam  2020   Là Gì?     4  Bệnh bạch hầu là gì        new
26954  Vietnam  2020   Là Gì?     5       Đông Lào là gì        new

Можно использовать метод `insert()`, чтобы указать, где должен быть новый столбец. У первого столбца индекс `0`:

In [75]:
df.insert(0, 'new_qwerty_column', 'qwerty')
df.head(5)

new_qwerty_column location  year         category  rank  query new_column
0            qwerty   Global  2001  Consumer Brands     1  Nokia        new
1            qwerty   Global  2001  Consumer Brands     2   Sony        new
2            qwerty   Global  2001  Consumer Brands     3    BMW        new
3            qwerty   Global  2001  Consumer Brands     4   Palm        new
4            qwerty   Global  2001  Consumer Brands     5  Adobe        new

### 3.3. Удаление строчек и столбцов

Удалим последнюю строчку под номером `26954` c помощью `drop()`:

In [76]:
df = pd.read_csv('https://raw.githubusercontent.com/tttdddnet/Python-Data-Journalism/main/part1/trends.csv')
df.tail(7)

location  year      category  rank                        query
26948  Vietnam  2020  Như Thế Nào?     4  Tuấn khỉ bị bắt như thế nào
26949  Vietnam  2020  Như Thế Nào?     5           Gấu đi như thế nào
26950  Vietnam  2020        Là Gì?     1           Virus Corona là gì
26951  Vietnam  2020        Là Gì?     2          Miễn thị thực là gì
26952  Vietnam  2020        Là Gì?     3            Đầu cắt moi là gì
26953  Vietnam  2020        Là Gì?     4          Bệnh bạch hầu là gì
26954  Vietnam  2020        Là Gì?     5               Đông Lào là gì

In [77]:
df = df.drop(26954, 0)
df.tail(7)

location  year      category  rank                             query
26947  Vietnam  2020  Như Thế Nào?     3  Bầu cử tổng thống mỹ như thế nào
26948  Vietnam  2020  Như Thế Nào?     4       Tuấn khỉ bị bắt như thế nào
26949  Vietnam  2020  Như Thế Nào?     5                Gấu đi như thế nào
26950  Vietnam  2020        Là Gì?     1                Virus Corona là gì
26951  Vietnam  2020        Là Gì?     2               Miễn thị thực là gì
26952  Vietnam  2020        Là Gì?     3                 Đầu cắt moi là gì
26953  Vietnam  2020        Là Gì?     4               Bệnh bạch hầu là gì

Здесь:  
`year` – название столбца  
`0` – в датафрейме строчка имеет `ось = 0`, столбец – `1`.  

Удалим столбец `year`:

In [78]:
df = df.drop('year', 1)
df.tail(5)

location      category  rank                query
26949  Vietnam  Như Thế Nào?     5   Gấu đi như thế nào
26950  Vietnam        Là Gì?     1   Virus Corona là gì
26951  Vietnam        Là Gì?     2  Miễn thị thực là gì
26952  Vietnam        Là Gì?     3    Đầu cắt moi là gì
26953  Vietnam        Là Gì?     4  Bệnh bạch hầu là gì

### 3.4. Переименование столбцов

Чтобы переименовать столбец используем `rename()`:

In [79]:
df = pd.read_csv('https://raw.githubusercontent.com/tttdddnet/Python-Data-Journalism/main/part1/trends.csv')
df.tail(3)

location  year category  rank                query
26952  Vietnam  2020   Là Gì?     3    Đầu cắt moi là gì
26953  Vietnam  2020   Là Gì?     4  Bệnh bạch hầu là gì
26954  Vietnam  2020   Là Gì?     5       Đông Lào là gì

In [80]:
df = df.rename(columns={'location': 'new_location'})
df.tail(3)

new_location  year category  rank                query
26952      Vietnam  2020   Là Gì?     3    Đầu cắt moi là gì
26953      Vietnam  2020   Là Gì?     4  Bệnh bạch hầu là gì
26954      Vietnam  2020   Là Gì?     5       Đông Lào là gì

Здесь:  
`location` – текущее название столбца.  
`new_location` – новое название столбца.

### 3.5. Замена символов в строке

С помощью `replace()` заменим в столбце `year` значения c 2020 на 2021:

In [81]:
df = pd.read_csv('https://raw.githubusercontent.com/tttdddnet/Python-Data-Journalism/main/part1/trends.csv')
df.tail(10)

location  year      category  rank                             query
26945  Vietnam  2020  Như Thế Nào?     1        Cúng giao thừa như thế nào
26946  Vietnam  2020  Như Thế Nào?     2      Vụ án Hồ Duy Hải như thế nào
26947  Vietnam  2020  Như Thế Nào?     3  Bầu cử tổng thống mỹ như thế nào
26948  Vietnam  2020  Như Thế Nào?     4       Tuấn khỉ bị bắt như thế nào
26949  Vietnam  2020  Như Thế Nào?     5                Gấu đi như thế nào
26950  Vietnam  2020        Là Gì?     1                Virus Corona là gì
26951  Vietnam  2020        Là Gì?     2               Miễn thị thực là gì
26952  Vietnam  2020        Là Gì?     3                 Đầu cắt moi là gì
26953  Vietnam  2020        Là Gì?     4               Bệnh bạch hầu là gì
26954  Vietnam  2020        Là Gì?     5                    Đông Lào là gì

In [82]:
df['year'] = df['year'].replace(to_replace=2020, value=2021)

# сокращенная запись без to_replace и value: 
# df['year'] = df['year'].replace(2020, 2021)

df['category'] = df['category'].replace(to_replace='Là Gì?', value='Hello, World.')
df.tail(10)

location  year       category  rank                             query
26945  Vietnam  2021   Như Thế Nào?     1        Cúng giao thừa như thế nào
26946  Vietnam  2021   Như Thế Nào?     2      Vụ án Hồ Duy Hải như thế nào
26947  Vietnam  2021   Như Thế Nào?     3  Bầu cử tổng thống mỹ như thế nào
26948  Vietnam  2021   Như Thế Nào?     4       Tuấn khỉ bị bắt như thế nào
26949  Vietnam  2021   Như Thế Nào?     5                Gấu đi như thế nào
26950  Vietnam  2021  Hello, World.     1                Virus Corona là gì
26951  Vietnam  2021  Hello, World.     2               Miễn thị thực là gì
26952  Vietnam  2021  Hello, World.     3                 Đầu cắt moi là gì
26953  Vietnam  2021  Hello, World.     4               Bệnh bạch hầu là gì
26954  Vietnam  2021  Hello, World.     5                    Đông Lào là gì

Здесь:  
`replace()` – заменяет значения.

`to_replace=2020` – значение, подлежащее замене.  
`value=2021` – то, на что меняем.  

Метод `replace()` выполнит замену в результате полного совпадения значения, то есть знак вопроса `?` просто так не удалить – нужно писать регулярное выражение (regex).
Быстрее воспользоваться методом `str.replace()`, который выполняет замену подстроки без regex. Заменим знак вопроса обоими методами.

Замена с помощью метода `replace()` и регулярных выражений:

In [83]:
df = pd.read_csv('https://raw.githubusercontent.com/tttdddnet/Python-Data-Journalism/main/part1/trends.csv')
df['category'] = df['category'].replace(r'[?]', '', regex=True)
df.tail(5)

location  year category  rank                query
26950  Vietnam  2020    Là Gì     1   Virus Corona là gì
26951  Vietnam  2020    Là Gì     2  Miễn thị thực là gì
26952  Vietnam  2020    Là Gì     3    Đầu cắt moi là gì
26953  Vietnam  2020    Là Gì     4  Bệnh bạch hầu là gì
26954  Vietnam  2020    Là Gì     5       Đông Lào là gì

Здесь:  
`r''` – обозначение регулярного выражения.   
`[]` – поиск только символа в скобках.  
`?` – поиск знака вопроса.  

Замена, используя метод `str.replace()`:

In [84]:
df = pd.read_csv('https://raw.githubusercontent.com/tttdddnet/Python-Data-Journalism/main/part1/trends.csv')
df['category'] = df['category'].str.replace('?', '', regex=False)
df.tail(5)

location  year category  rank                query
26950  Vietnam  2020    Là Gì     1   Virus Corona là gì
26951  Vietnam  2020    Là Gì     2  Miễn thị thực là gì
26952  Vietnam  2020    Là Gì     3    Đầu cắt moi là gì
26953  Vietnam  2020    Là Gì     4  Bệnh bạch hầu là gì
26954  Vietnam  2020    Là Gì     5       Đông Lào là gì

Здесь:  
`regex=False` – знак вопроса `?` используется в регулярных выражениях, поэтому их нужно отключить, присвоив параметру regex значение `False`. По умолчанию `True`.

Также можно заменить значения, используя метод `mask()`:

In [85]:
df['year'] = df['year'].mask(df['year'] == 2001, 2051)
df.head(3)

location  year         category  rank  query
0   Global  2051  Consumer Brands     1  Nokia
1   Global  2051  Consumer Brands     2   Sony
2   Global  2051  Consumer Brands     3    BMW

### 3.6. Суммирование элементов

Просуммируем элементы в столбце `year` с помощью `sum()`:

In [86]:
df = pd.read_csv('https://raw.githubusercontent.com/tttdddnet/Python-Data-Journalism/main/part1/trends.csv')
df['year'].sum()

54320885

### 3.7. Копирование датафрейма

Создадим глубокую копию (изменение в оригинале не влияют на копию и наоборот) датафрейма с помощью `copy()`:

In [87]:
df = pd.read_csv('https://raw.githubusercontent.com/tttdddnet/Python-Data-Journalism/main/part1/trends.csv')
df2 = df.copy

Здесь:  
`df.copy` – создает независимую копию датафрейма. Если нужна зависимая копия, то параметру `deep` присвоим значение `False`: `df.copy(deep=False)`. По умолчанию `True`.

## 4. Сортировка

Сейчас массив отсортирован по году, давайте отсортируем его по стране (`location`) в прямом алфавитном порядке с помощью `sort_values()`:

In [88]:
df = pd.read_csv('https://raw.githubusercontent.com/tttdddnet/Python-Data-Journalism/main/part1/trends.csv')
df.sort_values('location')

location  year    category  rank  \
5261   Argentina  2013  Tecnología     2   
16288  Argentina  2017        Cómo     4   
16287  Argentina  2017        Cómo     3   
16286  Argentina  2017        Cómo     2   
16285  Argentina  2017        Cómo     1   
...          ...   ...         ...   ...   
21361   Zimbabwe  2018   Questions     2   
21362   Zimbabwe  2018   Questions     3   
21363   Zimbabwe  2018   Questions     4   
21365   Zimbabwe  2018    Searches     1   
18808   Zimbabwe  2017    Searches     4   

                                           query  
5261                              Nokia Asha 311  
16288                           Cómo hacer slime  
16287  Cómo saber si estoy afiliado a un partido  
16286                Cómo hacer huevos de pascua  
16285                      Cómo saber dónde voto  
...                                          ...  
21361                                How to kiss  
21362                               What is love  
21363         How to write an application letter  
21365                         World Cup Fixtures  
18808                              Action movies  

[26955 rows x 5 columns]

и в обратном алфавитном порядке:

In [89]:
df.sort_values('location', ascending=False)

location  year   category  rank                               query
21367   Zimbabwe  2018   Searches     3                   Morgan Tsvangirai
21361   Zimbabwe  2018  Questions     2                         How to kiss
21363   Zimbabwe  2018  Questions     4  How to write an application letter
21364   Zimbabwe  2018  Questions     5               What is communication
21365   Zimbabwe  2018   Searches     1                  World Cup Fixtures
...          ...   ...        ...   ...                                 ...
24061  Argentina  2020   Personas     2                         Kobe Bryant
24062  Argentina  2020   Personas     3                         Naya Rivera
24063  Argentina  2020   Personas     4               Luis Alberto Spinetta
24064  Argentina  2020   Personas     5                        Elsa Serrano
5245   Argentina  2013  Películas     1                 Mi Villano Favorito

[26955 rows x 5 columns]

Здесь:  
`ascending=False` – сортировка в обратном алфавитном порядке. По умолчанию `True`.

Отсортируем по двум столбцам: `rank` и `category`:

In [90]:
df.sort_values(['rank', 'category'])

location  year               category  rank                 query
7280         Spain  2013              "Cómo..."     1   Cómo pagar Whatsapp
9000       Ireland  2014      "How To" Searches     1           How to draw
7285         Spain  2013            "Qué es..."     1        Qué es Airdrop
16385      Austria  2017       "Was...?" Fragen     1          Was ist G20?
18965      Austria  2018       "Was...?" Fragen     1  Was ist los in Wien?
...            ...   ...                    ...   ...                   ...
1359   South Korea  2008          최다 검색어 TOP 10     5                 Korea
1364   South Korea  2008  최다 검색어 TOP 10 (포털 제외)     5                    검색
26274  South Korea  2020                  코로나19     5                   마스크
12459  South Korea  2015                     키즈     5             마다가스카의 펭귄
7279   South Korea  2013                 화제의 인물     5                  크레용팝

[26955 rows x 5 columns]

В этом случаем приоритет отдается столбцу `rank`, так как он записан первым.

In [91]:
df.sort_values(['rank', 'category'], ascending=[False, False])

location  year               category  rank                 query
7279   South Korea  2013                 화제의 인물     5                  크레용팝
12459  South Korea  2015                     키즈     5             마다가스카의 펭귄
26274  South Korea  2020                  코로나19     5                   마스크
1364   South Korea  2008  최다 검색어 TOP 10 (포털 제외)     5                    검색
1359   South Korea  2008          최다 검색어 TOP 10     5                 Korea
...            ...   ...                    ...   ...                   ...
16385      Austria  2017       "Was...?" Fragen     1          Was ist G20?
18965      Austria  2018       "Was...?" Fragen     1  Was ist los in Wien?
7285         Spain  2013            "Qué es..."     1        Qué es Airdrop
9000       Ireland  2014      "How To" Searches     1           How to draw
7280         Spain  2013              "Cómo..."     1   Cómo pagar Whatsapp

[26955 rows x 5 columns]

Здесь:  
`ascending=[False, False]` – сортировка в обратном порядке по обоим столбцам.

## 5. Очистка данных

### 5.1. Изменение типа данных

Чтобы изменить тип данных воспользуемся методом `astype()`:

In [92]:
df = pd.read_csv('https://raw.githubusercontent.com/tttdddnet/Python-Data-Journalism/main/part1/trends.csv')
df.year.dtype

dtype('int64')

In [93]:
df['year'] = df['year'].astype(str)
df.year.dtype

dtype('O')

In [94]:
df['year'] = df['year'].astype(float)
df.year.dtype

dtype('float64')

In [95]:
df['year'] = df['year'].astype('int32')
df.year.dtype

dtype('int32')

In [96]:
df['year'] = df['year'].astype('int64')
df.year.dtype

dtype('int64')

In [97]:
df['year'] = pd.to_numeric(df['year'])
df.year.dtype

dtype('int64')

Здесь:  
`astype()` – меняет тип переменной на строчный `O`; на число с плавающей точкой `float`; на целые числа `int32` и `int64`.  
  
`to_numeric()` – меняет типа переменной на `int64`.

### 5.2. Удаление и замена NaN-значений

Добавим в начало датафрейма строчку с отсутствующими значениями `NaN`:

In [98]:
import numpy as np

df = pd.read_csv('https://raw.githubusercontent.com/tttdddnet/Python-Data-Journalism/main/part1/trends.csv')

df.loc[0] = ['Russia', np.nan, 'Music', np.nan, 'Artist Name']
df = df.append(new_row, ignore_index=True)
df.head(5)

location    year         category  rank        query
0   Russia     NaN            Music   NaN  Artist Name
1   Global  2001.0  Consumer Brands   2.0         Sony
2   Global  2001.0  Consumer Brands   3.0          BMW
3   Global  2001.0  Consumer Brands   4.0         Palm
4   Global  2001.0  Consumer Brands   5.0        Adobe

Узнаем, какие элементы имеют значение `NaN` с помощью метода `isnull()`:

In [99]:
df.isnull().head(3)

location   year  category   rank  query
0     False   True     False   True  False
1     False  False     False  False  False
2     False  False     False  False  False

и наоборот методом `notnull()`:

In [100]:
df.notnull().head(3)

location   year  category   rank  query
0      True  False      True  False   True
1      True   True      True   True   True
2      True   True      True   True   True

Удалим строчки с отсутствующими значениями, используя `dropna()`:

In [103]:
df.dropna().head(3)

location    year         category  rank query
1   Global  2001.0  Consumer Brands   2.0  Sony
2   Global  2001.0  Consumer Brands   3.0   BMW
3   Global  2001.0  Consumer Brands   4.0  Palm

Удалим столбцы с отсутствующими значениями:

In [104]:
df = df.append(new_row, ignore_index=True)
df.dropna(axis=1).head(3)

location         category        query
0   Russia            Music  Artist Name
1   Global  Consumer Brands         Sony
2   Global  Consumer Brands          BMW

Заменим `NaN` на `abcd` методом `fillna()`:

In [105]:
df = df.append(new_row, ignore_index=True)
df.fillna('abcd').head(5)

location    year         category  rank        query
0   Russia    abcd            Music  abcd  Artist Name
1   Global  2001.0  Consumer Brands   2.0         Sony
2   Global  2001.0  Consumer Brands   3.0          BMW
3   Global  2001.0  Consumer Brands   4.0         Palm
4   Global  2001.0  Consumer Brands   5.0        Adobe

### 5.3. Удаление лишних пробелов

Добавим строчку, в которой есть дефектные элементы с лишними пробелами в конце и в начале:

In [106]:
df = pd.read_csv('https://raw.githubusercontent.com/tttdddnet/Python-Data-Journalism/main/part1/trends.csv')
new_row = {'location': '  Russia     ', 'year': 2021, 'category': 'Music', 'rank': 1, 'query': ' Artist Name  '}

df = df.append(new_row, ignore_index=True)
print(df.tail(2))

            location  year category  rank           query
26954        Vietnam  2020   Là Gì?     5  Đông Lào là gì
26955    Russia       2021    Music     1   Artist Name  


Применим метод `strip()`, который удаляет символы (по умолчанию – пробел) с левого и правого края строки.

In [107]:
df['location'] = df['location'].map(str.strip)
df['query'] = df['query'].map(str.strip)
print(df.tail(2))

      location  year category  rank           query
26954  Vietnam  2020   Là Gì?     5  Đông Lào là gì
26955   Russia  2021    Music     1     Artist Name


Здесь:  
`map()` – применяет к каждому элементу столбцов `location` и `query` метод `strip()`.

Напишем цикл, который проходится по всем элементам датафрейма и удаляет у строчек лишние пробелы слева и справа:

In [108]:
df = pd.read_csv('https://raw.githubusercontent.com/tttdddnet/Python-Data-Journalism/main/part1/trends.csv')
new_row1 = {'location': '  Russia     ', 'year': 2021, 'category': 'Music', 'rank': 1, 'query': ' Artist Name  '}
new_row2 = {'location': ' Russia       ', 'year': 2021, 'category': 'Music  ', 'rank': 1, 'query': ' Artist Name '}

df = df.append(new_row1, ignore_index=True)
df = df.append(new_row2, ignore_index=True)
print(df.tail(3))

             location  year category  rank           query
26954         Vietnam  2020   Là Gì?     5  Đông Lào là gì
26955     Russia       2021    Music     1   Artist Name  
26956   Russia         2021  Music       1    Artist Name 


In [109]:
for column in df.columns:
    if df[column].dtype == object:
        df[column] = df[column].map(str.strip)

print(df.tail(3))

      location  year category  rank           query
26954  Vietnam  2020   Là Gì?     5  Đông Lào là gì
26955   Russia  2021    Music     1     Artist Name
26956   Russia  2021    Music     1     Artist Name


Здесь:  
`new_row1` и `new_row2` – добавляемые строчки.
`if df[column].dtype == object` – проверяем, является ли переменная объектом.

Проверить тип можно также через модуль `is_string_dtype`:

In [110]:
df = pd.read_csv('https://raw.githubusercontent.com/tttdddnet/Python-Data-Journalism/main/part1/trends.csv')
new_row1 = {'location': '  Russia     ', 'year': 2021, 'category': 'Music', 'rank': 1, 'query': ' Artist Name  '}
new_row2 = {'location': ' Russia       ', 'year': 2021, 'category': 'Music  ', 'rank': 1, 'query': ' Artist Name '}

df = df.append(new_row1, ignore_index=True)
df = df.append(new_row2, ignore_index=True)
print(df.tail(3))

             location  year category  rank           query
26954         Vietnam  2020   Là Gì?     5  Đông Lào là gì
26955     Russia       2021    Music     1   Artist Name  
26956   Russia         2021  Music       1    Artist Name 


In [111]:
from pandas.api.types import is_string_dtype

for column in df.columns:
    if is_string_dtype(df[column].dtype):
        df[column] = df[column].map(str.strip)

print(df.tail(3))

      location  year category  rank           query
26954  Vietnam  2020   Là Gì?     5  Đông Lào là gì
26955   Russia  2021    Music     1     Artist Name
26956   Russia  2021    Music     1     Artist Name


Здесь:  
`is_string_dtype()` – проверяет, является ли элемент строкой и возвращает `True` или `False`.  
Проверить на наличие числа можно через `is_numeric_dtype()`.

### 5.4. Обработка дат

Откроем сокращенный датасет ([оригинал](https://www.kaggle.com/jsphyg/weather-dataset-rattle-package)) c данными о погоде в Австралии:

In [148]:
df = pd.read_csv('https://raw.githubusercontent.com/tttdddnet/Python-Data-Journalism/main/part1/weatherAUS.csv')
df

Date Location  MinTemp  MaxTemp  Rainfall  Evaporation  Sunshine  \
0  2008-12-01   Albury     13.4     22.9       0.6          NaN       NaN   
1  2008-12-02   Albury      7.4     25.1       0.0          NaN       NaN   
2  2008-12-03   Albury     12.9     25.7       0.0          NaN       NaN   
3  2008-12-04   Albury      9.2     28.0       0.0          NaN       NaN   
4  2008-12-05   Albury     17.5     32.3       1.0          NaN       NaN   
5  2008-12-06   Albury     14.6     29.7       0.2          NaN       NaN   
6  2008-12-07   Albury     14.3     25.0       0.0          NaN       NaN   
7  2008-12-08   Albury      7.7     26.7       0.0          NaN       NaN   
8  2008-12-09   Albury      9.7     31.9       0.0          NaN       NaN   

  WindGustDir  WindGustSpeed WindDir9am  ... Humidity9am  Humidity3pm  \
0           W             44          W  ...          71           22   
1         WNW             44        NNW  ...          44           25   
2         WSW             46          W  ...          38           30   
3          NE             24         SE  ...          45           16   
4           W             41        ENE  ...          82           33   
5         WNW             56          W  ...          55           23   
6           W             50         SW  ...          49           19   
7           W             35        SSE  ...          48           19   
8         NNW             80         SE  ...          42            9   

   Pressure9am  Pressure3pm  Cloud9am  Cloud3pm  Temp9am  Temp3pm  RainToday  \
0       1007.7       1007.1       8.0       NaN     16.9     21.8         No   
1       1010.6       1007.8       NaN       NaN     17.2     24.3         No   
2       1007.6       1008.7       NaN       2.0     21.0     23.2         No   
3       1017.6       1012.8       NaN       NaN     18.1     26.5         No   
4       1010.8       1006.0       7.0       8.0     17.8     29.7         No   
5       1009.2       1005.4       NaN       NaN     20.6     28.9         No   
6       1009.6       1008.2       1.0       NaN     18.1     24.6         No   
7       1013.4       1010.1       NaN       NaN     16.3     25.5         No   
8       1008.9       1003.6       NaN       NaN     18.3     30.2         No   

   RainTomorrow  
0            No  
1            No  
2            No  
3            No  
4            No  
5            No  
6            No  
7            No  
8           Yes  

[9 rows x 23 columns]

Узнаем типы данных:

In [149]:
df.dtypes

Date              object
Location          object
MinTemp          float64
MaxTemp          float64
Rainfall         float64
Evaporation      float64
Sunshine         float64
WindGustDir       object
WindGustSpeed      int64
WindDir9am        object
WindDir3pm        object
WindSpeed9am       int64
WindSpeed3pm       int64
Humidity9am        int64
Humidity3pm        int64
Pressure9am      float64
Pressure3pm      float64
Cloud9am         float64
Cloud3pm         float64
Temp9am          float64
Temp3pm          float64
RainToday         object
RainTomorrow      object
dtype: object

Как видно, элементы столбца `Date` – строчки. Поменяем их тип на дату методом `to_datetime()`:

In [150]:
df['Date'] = pd.to_datetime(df['Date'])
df.dtypes

Date             datetime64[ns]
Location                 object
MinTemp                 float64
MaxTemp                 float64
Rainfall                float64
Evaporation             float64
Sunshine                float64
WindGustDir              object
WindGustSpeed             int64
WindDir9am               object
WindDir3pm               object
WindSpeed9am              int64
WindSpeed3pm              int64
Humidity9am               int64
Humidity3pm               int64
Pressure9am             float64
Pressure3pm             float64
Cloud9am                float64
Cloud3pm                float64
Temp9am                 float64
Temp3pm                 float64
RainToday                object
RainTomorrow             object
dtype: object

Добавим отдельно столбцы год, месяц, день из столбца `Date` методом `DatetimeIndex()`:

In [115]:
df['Year'] = pd.DatetimeIndex(df['Date']).year
df['Month'] = pd.DatetimeIndex(df['Date']).month
df['Day'] = pd.DatetimeIndex(df['Date']).weekday
df.head(10)

Date Location  MinTemp  MaxTemp  Rainfall  Evaporation  Sunshine  \
0 2008-12-01   Albury     13.4     22.9       0.6          NaN       NaN   
1 2008-12-02   Albury      7.4     25.1       0.0          NaN       NaN   
2 2008-12-03   Albury     12.9     25.7       0.0          NaN       NaN   
3 2008-12-04   Albury      9.2     28.0       0.0          NaN       NaN   
4 2008-12-05   Albury     17.5     32.3       1.0          NaN       NaN   
5 2008-12-06   Albury     14.6     29.7       0.2          NaN       NaN   
6 2008-12-07   Albury     14.3     25.0       0.0          NaN       NaN   
7 2008-12-08   Albury      7.7     26.7       0.0          NaN       NaN   
8 2008-12-09   Albury      9.7     31.9       0.0          NaN       NaN   

  WindGustDir  WindGustSpeed WindDir9am  ... Pressure3pm  Cloud9am  Cloud3pm  \
0           W             44          W  ...      1007.1       8.0       NaN   
1         WNW             44        NNW  ...      1007.8       NaN       NaN   
2         WSW             46          W  ...      1008.7       NaN       2.0   
3          NE             24         SE  ...      1012.8       NaN       NaN   
4           W             41        ENE  ...      1006.0       7.0       8.0   
5         WNW             56          W  ...      1005.4       NaN       NaN   
6           W             50         SW  ...      1008.2       1.0       NaN   
7           W             35        SSE  ...      1010.1       NaN       NaN   
8         NNW             80         SE  ...      1003.6       NaN       NaN   

   Temp9am  Temp3pm  RainToday  RainTomorrow  Year  Month  Day  
0     16.9     21.8         No            No  2008     12    0  
1     17.2     24.3         No            No  2008     12    1  
2     21.0     23.2         No            No  2008     12    2  
3     18.1     26.5         No            No  2008     12    3  
4     17.8     29.7         No            No  2008     12    4  
5     20.6     28.9         No            No  2008     12    5  
6     18.1     24.6         No            No  2008     12    6  
7     16.3     25.5         No            No  2008     12    0  
8     18.3     30.2         No           Yes  2008     12    1  

[9 rows x 26 columns]

Обратим внимание, что `DatetimeIndex.weekday` записывает не дату а индексы дней от `0` до `6`: понедельник имеет значение `0`, воскресенье – `6`.

Другой способ записи:

In [116]:
df['Year2'] = df['Date'].dt.year
df['Month2'] = df['Date'].dt.month
df['Day2'] = df['Date'].dt.day
df.head(10)

Date Location  MinTemp  MaxTemp  Rainfall  Evaporation  Sunshine  \
0 2008-12-01   Albury     13.4     22.9       0.6          NaN       NaN   
1 2008-12-02   Albury      7.4     25.1       0.0          NaN       NaN   
2 2008-12-03   Albury     12.9     25.7       0.0          NaN       NaN   
3 2008-12-04   Albury      9.2     28.0       0.0          NaN       NaN   
4 2008-12-05   Albury     17.5     32.3       1.0          NaN       NaN   
5 2008-12-06   Albury     14.6     29.7       0.2          NaN       NaN   
6 2008-12-07   Albury     14.3     25.0       0.0          NaN       NaN   
7 2008-12-08   Albury      7.7     26.7       0.0          NaN       NaN   
8 2008-12-09   Albury      9.7     31.9       0.0          NaN       NaN   

  WindGustDir  WindGustSpeed WindDir9am  ... Temp9am  Temp3pm  RainToday  \
0           W             44          W  ...    16.9     21.8         No   
1         WNW             44        NNW  ...    17.2     24.3         No   
2         WSW             46          W  ...    21.0     23.2         No   
3          NE             24         SE  ...    18.1     26.5         No   
4           W             41        ENE  ...    17.8     29.7         No   
5         WNW             56          W  ...    20.6     28.9         No   
6           W             50         SW  ...    18.1     24.6         No   
7           W             35        SSE  ...    16.3     25.5         No   
8         NNW             80         SE  ...    18.3     30.2         No   

   RainTomorrow  Year  Month  Day  Year2  Month2  Day2  
0            No  2008     12    0   2008      12     1  
1            No  2008     12    1   2008      12     2  
2            No  2008     12    2   2008      12     3  
3            No  2008     12    3   2008      12     4  
4            No  2008     12    4   2008      12     5  
5            No  2008     12    5   2008      12     6  
6            No  2008     12    6   2008      12     7  
7            No  2008     12    0   2008      12     8  
8           Yes  2008     12    1   2008      12     9  

[9 rows x 29 columns]

## 6. Фильтрация и поиск

С помощью оператора сравнения `>=` получим все поисковые запросы c 2018 года включительно до 2020 года:

In [117]:
import pandas as pd
df = pd.read_csv('https://raw.githubusercontent.com/tttdddnet/Python-Data-Journalism/main/part1/trends.csv')
df[df['year'] >= 2018 ]

location  year  category  rank                query
18810   Global  2018  Searches     1            World Cup
18811   Global  2018  Searches     2               Avicii
18812   Global  2018  Searches     3           Mac Miller
18813   Global  2018  Searches     4             Stan Lee
18814   Global  2018  Searches     5        Black Panther
...        ...   ...       ...   ...                  ...
26950  Vietnam  2020    Là Gì?     1   Virus Corona là gì
26951  Vietnam  2020    Là Gì?     2  Miễn thị thực là gì
26952  Vietnam  2020    Là Gì?     3    Đầu cắt moi là gì
26953  Vietnam  2020    Là Gì?     4  Bệnh bạch hầu là gì
26954  Vietnam  2020    Là Gì?     5       Đông Lào là gì

[8145 rows x 5 columns]

Более сложные выражения записываются с побитовыми операторами. В данном случае с оператором `&` (`И`, `AND`):

In [118]:
df.loc[(df['year'] >= 2015) & (df['year'] < 2017) & (df['location'] == 'Global')]

location  year  category  rank            query
10370   Global  2015  Searches     1       Lamar Odom
10371   Global  2015  Searches     2    Charlie Hebdo
10372   Global  2015  Searches     3          Agar.io
10373   Global  2015  Searches     4   Jurassic World
10374   Global  2015  Searches     5            Paris
...        ...   ...       ...   ...              ...
13200   Global  2016  TV Shows     1  Stranger Things
13201   Global  2016  TV Shows     2        Westworld
13202   Global  2016  TV Shows     3        Luke Cage
13203   Global  2016  TV Shows     4  Game of Thrones
13204   Global  2016  TV Shows     5     Black Mirror

[90 rows x 5 columns]

### 6.1. Метод startswith()


Проверим, совпадает ли начало каждого элемента строки с шаблоном `G`, используя метод `startwith()`:

In [119]:
import pandas as pd
df = pd.read_csv('https://raw.githubusercontent.com/tttdddnet/Python-Data-Journalism/main/part1/trends.csv')
df['location_startswith'] = df['location'].str.startswith('G')
df.head(3)

location  year         category  rank  query  location_startswith
0   Global  2001  Consumer Brands     1  Nokia                 True
1   Global  2001  Consumer Brands     2   Sony                 True
2   Global  2001  Consumer Brands     3    BMW                 True

### 6.2. Метод endswith()

Метод `endswith()` проверяет совпадает ли конец каждого элемента строки с шаблоном `a`:

In [120]:
df['location_endswith'] = df['location'].str.endswith('a')
df.head(3)

location  year         category  rank  query  location_startswith  \
0   Global  2001  Consumer Brands     1  Nokia                 True   
1   Global  2001  Consumer Brands     2   Sony                 True   
2   Global  2001  Consumer Brands     3    BMW                 True   

   location_endswith  
0              False  
1              False  
2              False

Методы `startwith()` и `endswith()` не поддерживают регулярные выражения. Для более точного поиска рассмотрим следующие методы.

### 6.3. Метод match()

Метод `match()` определяет, начинается ли каждая строка с шаблона, возвращая булево значение. Выведем все запросы о моделях `Nokia`:

In [121]:
df = pd.read_csv('https://raw.githubusercontent.com/tttdddnet/Python-Data-Journalism/main/part1/trends.csv')
df['query'].str.match('Nokia', case=True)

0         True
1        False
2        False
3        False
4        False
         ...  
26950    False
26951    False
26952    False
26953    False
26954    False
Name: query, Length: 26955, dtype: bool

Чтобы вывести больше информации заключим предыдущую запись в `df[]`:

In [33]:
df[df['query'].str.match('Nokia', case=True)]

location  year                                        category  \
0            Global  2001                                 Consumer Brands   
29           Global  2001                                    New Products   
67           Global  2002                                 Consumer Brands   
265           Italy  2003                                  Popular Brands   
999         Finland  2008                                   Top Companies   
1062      Hong Kong  2008                                            數碼產品   
1553         Global  2009  Looking to Buy Online: Fastest Rising Searches   
1957         Global  2010                            Consumer Electronics   
1959         Global  2010                            Consumer Electronics   
3031      Argentina  2012                                        Shopping   
3032      Argentina  2012                                        Shopping   
3489        Finland  2012                                  Mobile Devices   
3583        Germany  2012                                         Gadgets   
4524   South Africa  2012                                Gadgets & Phones   
4774         Uganda  2012                                Gadgets & Phones   
5261      Argentina  2013                                      Tecnología   
5262      Argentina  2013                                      Tecnología   
7996        Vietnam  2013                                      Công nghệ    
8030        Vietnam  2013                                    Thương Hiệu    
8821          Ghana  2014                                   Mobile Phones   
8823          Ghana  2014                                   Mobile Phones   
8961      Indonesia  2014                                          Gadget   
9351        Nigeria  2014                                   Mobile Phones   
9999       Thailand  2014                            แก็ดเจ็ตและเทคโนโลยี   
10367       Vietnam  2014                                      Điện thoại   
17103     Hong Kong  2017                                       熱爆 Gadget   
17104     Hong Kong  2017                                       熱爆 Gadget   

       rank            query  
0         1            Nokia  
29        5       Nokia 5510  
67        3            Nokia  
265       1            Nokia  
999       5            Nokia  
1062      3            Nokia  
1553      4        Nokia N97  
1957      3       Nokia 5530  
1959      5       Nokia N900  
3031      2        Nokia 500  
3032      3  Nokia lumia 900  
3489      5        Nokia 500  
3583      4  Nokia Lumia 920  
4524      5         Nokia n9  
4774      5   Nokia Asha 303  
5261      2   Nokia Asha 311  
5262      3  Nokia Lumia 620  
7996      2  Nokia Lumia 520  
8030      1            Nokia  
8821      2         Nokia XL  
8823      4          Nokia X  
8961      2          Nokia X  
9351      2          Nokia X  
9999      5          Nokia X  
10367     3          Nokia X  
17103     4          Nokia 8  
17104     5          Nokia 6

В запросах попадается название компании `Nokia` без модели, например, `Nokia 8`. Уточним запрос, воспользовавшись регулярным выражением:

In [122]:
import re
df[df['query'].str.match(r'\b..[k][i][a]\s.', case=True)]

location  year                                        category  \
29           Global  2001                                    New Products   
1553         Global  2009  Looking to Buy Online: Fastest Rising Searches   
1957         Global  2010                            Consumer Electronics   
1959         Global  2010                            Consumer Electronics   
3031      Argentina  2012                                        Shopping   
3032      Argentina  2012                                        Shopping   
3489        Finland  2012                                  Mobile Devices   
3583        Germany  2012                                         Gadgets   
4524   South Africa  2012                                Gadgets & Phones   
4774         Uganda  2012                                Gadgets & Phones   
5261      Argentina  2013                                      Tecnología   
5262      Argentina  2013                                      Tecnología   
7996        Vietnam  2013                                      Công nghệ    
8821          Ghana  2014                                   Mobile Phones   
8823          Ghana  2014                                   Mobile Phones   
8961      Indonesia  2014                                          Gadget   
9351        Nigeria  2014                                   Mobile Phones   
9999       Thailand  2014                            แก็ดเจ็ตและเทคโนโลยี   
10367       Vietnam  2014                                      Điện thoại   
17103     Hong Kong  2017                                       熱爆 Gadget   
17104     Hong Kong  2017                                       熱爆 Gadget   

       rank            query  
29        5       Nokia 5510  
1553      4        Nokia N97  
1957      3       Nokia 5530  
1959      5       Nokia N900  
3031      2        Nokia 500  
3032      3  Nokia lumia 900  
3489      5        Nokia 500  
3583      4  Nokia Lumia 920  
4524      5         Nokia n9  
4774      5   Nokia Asha 303  
5261      2   Nokia Asha 311  
5262      3  Nokia Lumia 620  
7996      2  Nokia Lumia 520  
8821      2         Nokia XL  
8823      4          Nokia X  
8961      2          Nokia X  
9351      2          Nokia X  
9999      5          Nokia X  
10367     3          Nokia X  
17103     4          Nokia 8  
17104     5          Nokia 6

Здесь:  
`import re` – импорт библиотеки решулярных выражений.  
`match()` – ищет совпадение в начале строки.  
`\b` – обозначает границу слова. В нашем случае: слева – ничего, справа – буква..  
`.` – любой один символ, кроме символа переноса строки.    
`[k][i][a]` – только символы `k`, `i` и `a` соответственно.  
`\s` – любой пробельный символ.  
`case=True` – чувствителен к регистру.  

### 6.4. Метод findall()

Метод `findall` возвращает совпадения шаблонов:

In [123]:
df['findall'] = df['query'].str.findall(r'\b..[k][i][a]\s.', flags=re.IGNORECASE).transform(''.join)
df.loc[df['findall'] != '']

location  year                                        category  \
29           Global  2001                                    New Products   
1553         Global  2009  Looking to Buy Online: Fastest Rising Searches   
1957         Global  2010                            Consumer Electronics   
1959         Global  2010                            Consumer Electronics   
3031      Argentina  2012                                        Shopping   
3032      Argentina  2012                                        Shopping   
3489        Finland  2012                                  Mobile Devices   
3583        Germany  2012                                         Gadgets   
4524   South Africa  2012                                Gadgets & Phones   
4774         Uganda  2012                                Gadgets & Phones   
5261      Argentina  2013                                      Tecnología   
5262      Argentina  2013                                      Tecnología   
7996        Vietnam  2013                                      Công nghệ    
8821          Ghana  2014                                   Mobile Phones   
8823          Ghana  2014                                   Mobile Phones   
8961      Indonesia  2014                                          Gadget   
9351        Nigeria  2014                                   Mobile Phones   
9999       Thailand  2014                            แก็ดเจ็ตและเทคโนโลยี   
10367       Vietnam  2014                                      Điện thoại   
17103     Hong Kong  2017                                       熱爆 Gadget   
17104     Hong Kong  2017                                       熱爆 Gadget   

       rank            query  findall  
29        5       Nokia 5510  Nokia 5  
1553      4        Nokia N97  Nokia N  
1957      3       Nokia 5530  Nokia 5  
1959      5       Nokia N900  Nokia N  
3031      2        Nokia 500  Nokia 5  
3032      3  Nokia lumia 900  Nokia l  
3489      5        Nokia 500  Nokia 5  
3583      4  Nokia Lumia 920  Nokia L  
4524      5         Nokia n9  Nokia n  
4774      5   Nokia Asha 303  Nokia A  
5261      2   Nokia Asha 311  Nokia A  
5262      3  Nokia Lumia 620  Nokia L  
7996      2  Nokia Lumia 520  Nokia L  
8821      2         Nokia XL  Nokia X  
8823      4          Nokia X  Nokia X  
8961      2          Nokia X  Nokia X  
9351      2          Nokia X  Nokia X  
9999      5          Nokia X  Nokia X  
10367     3          Nokia X  Nokia X  
17103     4          Nokia 8  Nokia 8  
17104     5          Nokia 6  Nokia 6

Здесь:  
`df['findall'] != '']` – отбирает все значения, за исключением пустой строчки.  
`flags=re.IGNORECASE` – игнорирует регистр при поиске.  
`transform(''.join)` – по умолчанию `findall()` заключает все запросы в квадратные скобки `[]`, что не очень удобно для последующей обработки. Поэтому мы заменили скобки на пустое место `''`.

Напишем цикл, который выводит отфильтрованные запросы:

In [124]:
for index, value in df['findall'].items():
    if value != '':
        print(f"Индекс: {index}, Значение: {df.iloc[index].query}")

Индекс: 29, Значение: Nokia 5510
Индекс: 1553, Значение: Nokia N97
Индекс: 1957, Значение: Nokia 5530
Индекс: 1959, Значение: Nokia N900
Индекс: 3031, Значение: Nokia 500
Индекс: 3032, Значение: Nokia lumia 900
Индекс: 3489, Значение: Nokia 500
Индекс: 3583, Значение: Nokia Lumia 920
Индекс: 4524, Значение: Nokia n9
Индекс: 4774, Значение: Nokia Asha 303
Индекс: 5261, Значение: Nokia Asha 311
Индекс: 5262, Значение: Nokia Lumia 620
Индекс: 7996, Значение: Nokia Lumia 520
Индекс: 8821, Значение: Nokia XL
Индекс: 8823, Значение: Nokia X
Индекс: 8961, Значение: Nokia X
Индекс: 9351, Значение: Nokia X
Индекс: 9999, Значение: Nokia X
Индекс: 10367, Значение: Nokia X
Индекс: 17103, Значение: Nokia 8
Индекс: 17104, Значение: Nokia 6


### 6.5. Метод str.contains()

Метод `str.contains()` в результате поиска возвращает булево значение:

In [125]:
df = pd.read_csv('https://raw.githubusercontent.com/tttdddnet/Python-Data-Journalism/main/part1/trends.csv')
df['query'].str.contains(r'\b..[k][i][a]\s.', regex=True)

0        False
1        False
2        False
3        False
4        False
         ...  
26950    False
26951    False
26952    False
26953    False
26954    False
Name: query, Length: 26955, dtype: bool

In [126]:
df[df['query'].str.contains(r'\b..[k][i][a]\s.', regex=True)]

location  year                                        category  \
29           Global  2001                                    New Products   
1553         Global  2009  Looking to Buy Online: Fastest Rising Searches   
1957         Global  2010                            Consumer Electronics   
1959         Global  2010                            Consumer Electronics   
3031      Argentina  2012                                        Shopping   
3032      Argentina  2012                                        Shopping   
3489        Finland  2012                                  Mobile Devices   
3583        Germany  2012                                         Gadgets   
4524   South Africa  2012                                Gadgets & Phones   
4774         Uganda  2012                                Gadgets & Phones   
5261      Argentina  2013                                      Tecnología   
5262      Argentina  2013                                      Tecnología   
7996        Vietnam  2013                                      Công nghệ    
8821          Ghana  2014                                   Mobile Phones   
8823          Ghana  2014                                   Mobile Phones   
8961      Indonesia  2014                                          Gadget   
9351        Nigeria  2014                                   Mobile Phones   
9999       Thailand  2014                            แก็ดเจ็ตและเทคโนโลยี   
10367       Vietnam  2014                                      Điện thoại   
17103     Hong Kong  2017                                       熱爆 Gadget   
17104     Hong Kong  2017                                       熱爆 Gadget   

       rank            query  
29        5       Nokia 5510  
1553      4        Nokia N97  
1957      3       Nokia 5530  
1959      5       Nokia N900  
3031      2        Nokia 500  
3032      3  Nokia lumia 900  
3489      5        Nokia 500  
3583      4  Nokia Lumia 920  
4524      5         Nokia n9  
4774      5   Nokia Asha 303  
5261      2   Nokia Asha 311  
5262      3  Nokia Lumia 620  
7996      2  Nokia Lumia 520  
8821      2         Nokia XL  
8823      4          Nokia X  
8961      2          Nokia X  
9351      2          Nokia X  
9999      5          Nokia X  
10367     3          Nokia X  
17103     4          Nokia 8  
17104     5          Nokia 6

### 6.6. Метод extractall()

Метод `extraclall()` вернет столбец с поисковым шаблоном. Результаты, не имеющие совпадений, отображаться не будут:

In [127]:
df = pd.read_csv('https://raw.githubusercontent.com/tttdddnet/Python-Data-Journalism/main/part1/trends.csv')
df['query'].str.extractall(r"(\b..[k][i][a]\s.)")

0
      match         
29    0      Nokia 5
1553  0      Nokia N
1957  0      Nokia 5
1959  0      Nokia N
3031  0      Nokia 5
3032  0      Nokia l
3489  0      Nokia 5
3583  0      Nokia L
4524  0      Nokia n
4774  0      Nokia A
5261  0      Nokia A
5262  0      Nokia L
7996  0      Nokia L
8821  0      Nokia X
8823  0      Nokia X
8961  0      Nokia X
9351  0      Nokia X
9999  0      Nokia X
10367 0      Nokia X
17103 0      Nokia 8
17104 0      Nokia 6

Присвоить название столбцу с результатами можно через следующую запись: `?P<название столбца>`. 

In [128]:
df['query'].str.extractall(r'(?P<model>\b..[k][i][a]\s.)')

model
      match         
29    0      Nokia 5
1553  0      Nokia N
1957  0      Nokia 5
1959  0      Nokia N
3031  0      Nokia 5
3032  0      Nokia l
3489  0      Nokia 5
3583  0      Nokia L
4524  0      Nokia n
4774  0      Nokia A
5261  0      Nokia A
5262  0      Nokia L
7996  0      Nokia L
8821  0      Nokia X
8823  0      Nokia X
8961  0      Nokia X
9351  0      Nokia X
9999  0      Nokia X
10367 0      Nokia X
17103 0      Nokia 8
17104 0      Nokia 6

Несколько фильтров в одном запросе:

In [129]:
df['query'].str.extractall(r'(?P<model>\b..[k][i][a]\s.)(?P<number>[5])')

model number
     match                
29   0      Nokia 5      5
1957 0      Nokia 5      5

### 6.7. Метод isin()

Узнаем, содержатся ли искомые значения в столбце методом `isin()` :

In [130]:
df = pd.read_csv('https://raw.githubusercontent.com/tttdddnet/Python-Data-Journalism/main/part1/trends.csv')
df['query'].isin(['Agar.io', 'Paris'])

0        False
1        False
2        False
3        False
4        False
         ...  
26950    False
26951    False
26952    False
26953    False
26954    False
Name: query, Length: 26955, dtype: bool

In [131]:
df.loc[df['query'].isin(['Agar.io', 'Paris'])]

location  year                               category  rank  \
105            Global  2002                    Travel Destinations     1   
2588        Singapore  2011                    Travel Destinations     4   
3761          Ireland  2012                    Travel Destinations     2   
4944   United Kingdom  2012                    Travel Destinations     5   
10372          Global  2015                               Searches     3   
10374          Global  2015                               Searches     5   
10376          Global  2015                            Global News     2   
10480       Australia  2015                     News Items of 2015     1   
10489         Austria  2015            Die Suchbegriffe des Jahres     5   
10491         Austria  2015                Schlagzeilen des Jahres     2   
10650          Brazil  2015                                  Games     1   
10783           Chile  2015                              Búsquedas     4   
10950         Denmark  2015                              Søgninger     1   
11174          France  2015                         Week end à ...     5   
11184         Germany  2015                           Suchbegriffe     5   
11188         Germany  2015                           Schlagzeilen     4   
11416       Indonesia  2015                                   Game     2   
11453         Ireland  2015                           News Moments     4   
11478         Ireland  2015                                Top Ten     4   
11755        Malaysia  2015                      Games (Permainan)     1   
11886     New Zealand  2015                     News Items of 2015     2   
12162          Poland  2015  Najszybciej zyskujące na popularności     3   
12308       Singapore  2015                     International News     4   
12544     Switzerland  2015                           Suchanfragen     5   
12806  United Kingdom  2015                           News Moments     2   
12839  United Kingdom  2015                                Top Ten     5   
13070   United States  2015                            Video Games     1   
13117         Vietnam  2015                            Tin Quốc tế     3   
13602           Chile  2016                            Videojuegos     3   
13643        Colombia  2016                                 Gaming     4   
13698      Costa Rica  2016                                 Gaming     4   
13916         Finland  2016                                  Pelit     2   
14633        Malaysia  2016                      Games (Permainan)     4   
14697          Mexico  2016                            Videojuegos     3   
14858          Panama  2016                                 Gaming     4   
16096   United States  2016                            Video Games     2   

         query  
105      Paris  
2588     Paris  
3761     Paris  
4944     Paris  
10372  Agar.io  
10374    Paris  
10376    Paris  
10480    Paris  
10489    Paris  
10491    Paris  
10650  Agar.io  
10783  Agar.io  
10950  Agar.io  
11174    Paris  
11184    Paris  
11188    Paris  
11416  Agar.io  
11453    Paris  
11478    Paris  
11755  Agar.io  
11886    Paris  
12162  Agar.io  
12308    Paris  
12544    Paris  
12806    Paris  
12839    Paris  
13070  Agar.io  
13117    Paris  
13602  Agar.io  
13643  Agar.io  
13698  Agar.io  
13916  Agar.io  
14633  Agar.io  
14697  Agar.io  
14858  Agar.io  
16096  Agar.io

## 7. Фильтрация по языку

### 7.1. Декодер

В датасете могут оказаться кракозябры: `סרטים`. Для борьбы с ними есть отличная библиотека `ftfy` (англ. fixes text for you – исправляет текст для тебя), которая также имеет [веб-реализацию](https://ftfy.now.sh/). Установим `ftfty`:

In [ ]:
!pip install ftfy

In [132]:
from ftfy import fix_encoding
df = pd.read_csv('https://raw.githubusercontent.com/tttdddnet/Python-Data-Journalism/main/part1/decode.csv')
df['decode'] = df['encode'].apply(fix_encoding)
df

encode  \
0                                     He's Justin❤   
1  Le Schtroumpf Docteur conseille g√¢teaux et ba...   
2                                    ✔ No problems   
3             дороге Из-под #футбол   
4  „Handwerk bringt dich überall hin“: Von der YO...   
5                                     Hi guys í ½í¸   
6                        hihi RT username: â˜ºí ½í¸˜   
7             Beta Haber: Hırsızı Büyü Korkuttu   
8  Kayanya laptopku error deh soalnya tiap mau ng...   

                                              decode  
0                                       He's Justin❤  
1  Le Schtroumpf Docteur conseille gâteaux et bai...  
2                                      ✔ No problems  
3                              дороге Из-под #футбол  
4  „Handwerk bringt dich überall hin“: Von der YO...  
5                                          Hi guys 😍  
6                               hihi RT username: ☺😘  
7                  Beta Haber: Hırsızı Büyü Korkuttu  
8  Kayanya laptopku error deh soalnya tiap mau ng...

Здесь:  
`apply(fix_encoding)` – метод `apply()` применяет к каждому элементу столбца `encode` функцию `fix_encoding`.

### 7.2. Определение языка

Библиотека `google_trans` оказалась с багами, поэтому воспользуемся ее работоспособным форком – `google_trans_new`. Установим `google_trans_new`:

In [ ]:
!pip install google_trans_new
!pip install requests

In [134]:
from google_trans_new import google_translator
import requests

df = pd.read_csv('https://raw.githubusercontent.com/tttdddnet/Python-Data-Journalism/main/part1/trends_lang.csv')
df['category']= df['category'].apply(fix_encoding)

detector = google_translator()  

for index, row in df['category'].items():
    try:
        df.loc[index, 'category language'] = detector.detect(row)[1]
        print(detector.detect(row)[1])
    except:
        df.loc[index, 'category language'] = 'Error'
        print('Error')

english
english
spanish
english
greek
spanish
korean
english
english
hebrew
hebrew
japanese
russian
russian
spanish
latvian
lithuanian
estonian
czech
danish
croatian
finnish
french
dutch


Здесь:  
`for index, row ...` – цикл по индексу и значению в столбце `category`.  

`try .... detector.detect(row)[1]` – определение языка. `0` – сокращенный вариант `en`, `1` – полный вариант `english`.  

`except ... 'Error'` – если по какой-то причине нам не удастся определить язык, в ячейку запишется значение `Error`.

In [135]:
df

location  year                           category  rank  \
0        Global  2001                    Consumer Brands     1   
1        Global  2001                    Consumer Brands     2   
2     Argentina  2015                             Qué...     2   
3        Global  2001                    Consumer Brands     4   
4        Greece  2016                       Διασημότητες     3   
5         Spain  2015     Los términos que más crecieron     2   
6   South Korea  2015                                조리법     5   
7        Global  2001                             Movies     1   
8        Global  2001                             Movies     2   
9        Israel  2015                              סרטים     1   
10       Israel  2015                              סרטים     2   
11        Japan  2016                                 映画     3   
12       Russia  2020   «Что такое...?»/«Что значит...?»     1   
13       Russia  2020   «Что такое...?»/«Что значит...?»     2   
14      Uruguay  2015                          Qué es...     3   
15       Latvia  2018        Visvairāk meklētie notikumi     4   
16    Lithuania  2018                             Filmai     3   
17      Estonia  2019  Kõige otsituimad märksõnad Eestis     1   
18      Czechia  2019                     Smutná loučení     2   
19      Denmark  2019           Mest trendende søgninger     2   
20      Croatia  2019                 Sportska događanja     1   
21      Finland  2019                       Puheenaiheet     5   
22       France  2020               Tendances de l'année     2   
23  Netherlands  2020                         Wat is...?     1   

                   query category language  
0                  Nokia           english  
1                   Sony           english  
2            Qué es ISIS           spanish  
3                   Palm           english  
4       Γιάννης Χαρούλης             greek  
5             Eurovisión           spanish  
6           파김치 맛있게 담는 법            korean  
7           Harry Potter           english  
8      Lord of the Rings           english  
9          מהיר ועצבני 7            hebrew  
10     50 גוונים של אפור            hebrew  
11                 ズートピア          japanese  
12              Пандемия           russian  
13           Коронавирус           russian  
14     Qué es un eclipse           spanish  
15          IKEA Latvijā           latvian  
16     Bohemian rhapsody        lithuanian  
17        Valimised 2019          estonian  
18         Cameron Boyce             czech  
19        Rasmus Paludan            danish  
20      Dinamo - Benfica          croatian  
21     Curly girl metodi           finnish  
22     Attestation Covid            french  
23  Wat is een lockdown?             dutch

Получим строки, содержащие язык `russian`:

In [136]:
df.loc[df['category language'] == "russian"]

location  year                          category  rank        query  \
12   Russia  2020  «Что такое...?»/«Что значит...?»     1     Пандемия   
13   Russia  2020  «Что такое...?»/«Что значит...?»     2  Коронавирус   

   category language  
12           russian  
13           russian

## 8. Статистические данные

### 8.1. Метод describe()

Для просмотра статистической сводки столбцов, содержащих численные значения, введем `df.describe()`.

In [137]:
df = pd.read_csv('https://raw.githubusercontent.com/tttdddnet/Python-Data-Journalism/main/part1/trends.csv')
df.describe()

year         rank
count  26955.000000  26955.00000
mean    2015.243369      3.00000
std        3.564683      1.41424
min     2001.000000      1.00000
25%     2013.000000      2.00000
50%     2016.000000      3.00000
75%     2018.000000      4.00000
max     2020.000000      5.00000

Здесь:  
`count` – считает количество записей в столбце без значения `NaN`. Отдельно можно вызвать через `df.count()`.  

`mean` – среднее значение `df.mean()`.  

`std` – стандартное отклонение `df.std()`.  

`min` – минимальное число в столбце `df.min()`.  

`25%`	– 25-й процентиль.  
`50%`	– 50-й процентиль.  
`75%`	– 75-й процентиль.  

`max` – максимальное число в столбце `df.max()`.

### 8.2. Подсчет повторяющихся значений

Подсчитаем количество элементов в столбце `query` методом `count()`:

In [138]:
df = pd.read_csv('https://raw.githubusercontent.com/tttdddnet/Python-Data-Journalism/main/part1/trends.csv')
df['query'].count()

26955

С помощью метода `value_counts` узнаем сколько раз значения повторяются в столбце:

In [139]:
df['query'].value_counts()

Paul Walker         84
Donald Trump        83
Facebook            62
Robin Williams      61
Whitney Houston     56
                    ..
長寿                   1
바운스                  1
Sidang MK            1
Kari Gjærum          1
Guillermo Dávila     1
Name: query, Length: 18431, dtype: int64

По умолчанию из результата исключаются NaN-значения. Чтобы отобразить их, поставим параметру `dropna` значение `False`.

In [140]:
df['query'].value_counts(dropna=False)

Paul Walker         84
Donald Trump        83
Facebook            62
Robin Williams      61
Whitney Houston     56
                    ..
長寿                   1
바운스                  1
Sidang MK            1
Kari Gjærum          1
Guillermo Dávila     1
Name: query, Length: 18431, dtype: int64

Создадим столбец `query_count`, в который запишем количество упоминаний элемента:

In [141]:
df['query_count'] = df.groupby('query')['query'].transform('count')
df.sample(5)

location  year          category  rank               query  query_count
6098    Israel  2013           מתכונים     4    עוגות לפסח פרווה            1
4837   Ukraine  2012            Фільми     3       Третій зайвий            1
16890  Finland  2017            Viihde     1  Tuntematon sotilas            1
23608   Taiwan  2019  快速竄升政治人物（不包含已故者）     4                 謝龍介            1
13796  Denmark  2016      Begivenheder     2                  OL            2

Здесь:  
`groupby()` – используется для разделения данных на группы по критериям. Подробнее [здесь](https://pandas.pydata.org/pandas-docs/stable/user_guide/groupby.html).  

`transform('count')` – считает сколько раз повторяется значение в столбце `query`.

Выведем уникальные значения в столбце `location` через метод `unique()`:

In [142]:
df['location'].unique()

array(['Global', 'France', 'Germany', 'United Kingdom', 'Australia',
       'Canada', 'Italy', 'Netherlands', 'Spain', 'United States',
       'Argentina', 'Austria', 'Belgium', 'Brazil', 'Chile', 'China',
       'Colombia', 'Czechia', 'Denmark', 'Finland', 'Hong Kong', 'India',
       'Malaysia', 'Mexico', 'New Zealand', 'Philippines', 'Poland',
       'Russia', 'Singapore', 'South Africa', 'South Korea', 'Sweden',
       'Switzerland', 'Taiwan', 'Thailand', 'United Arab Emirates',
       'Costa Rica', 'Croatia', 'Dominican Republic', 'Ecuador',
       'El Salvador', 'Guatemala', 'Honduras', 'Japan', 'Kenya',
       'Nigeria', 'Panama', 'Peru', 'Egypt', 'Hungary', 'Ireland',
       'Israel', 'Norway', 'Portugal', 'Romania', 'Saudi Arabia',
       'Serbia', 'Slovakia', 'Turkey', 'Ukraine', 'Ghana', 'Indonesia',
       'Senegal', 'Uganda', 'Vietnam', 'Bangladesh', 'Bulgaria',
       'Estonia', 'Latvia', 'Lithuania', 'Pakistan', 'Puerto Rico',
       'Slovenia', 'Uruguay', 'Venezuela', '

### 9. Сохранение датафрейма

Сохраним датафрейм в csv-файл:

In [143]:
df = pd.read_csv('https://raw.githubusercontent.com/tttdddnet/Python-Data-Journalism/main/part1/trends.csv')
save_file = df[['location', 'year', 'category', 'rank', 'query']]
save_file.to_csv('trends_saved.csv', encoding='utf8')

Здесь:  
`save_file` – столбцы датафрейма.  
`to_csv()` – сохранить датафрейм в файл формата `.csv`.

Откроем сохраненный файл:

In [144]:
df = pd.read_csv('https://raw.githubusercontent.com/tttdddnet/Python-Data-Journalism/main/part1/trends_saved.csv')
df.head(3)

Unnamed: 0 location  year         category  rank  query
0           0   Global  2001  Consumer Brands     1  Nokia
1           1   Global  2001  Consumer Brands     2   Sony
2           2   Global  2001  Consumer Brands     3    BMW

После сохранения файла у нас повявился столбец `Unnamed: 0`. Он нам не нужен, потому удалим его:

In [145]:
df = df.drop('Unnamed: 0', 1)
df.head(3)

location  year         category  rank  query
0   Global  2001  Consumer Brands     1  Nokia
1   Global  2001  Consumer Brands     2   Sony
2   Global  2001  Consumer Brands     3    BMW